File: packages/amplify_core/lib/src/types/storage/storage_path.dart
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import 'package:amplify_core/src/types/storage/storage_path_from_identity_id.dart';
import 'package:meta/meta.dart';

/// The full remote path of a storage item.
///
/// ### Examples
/// #### Static Path
/// {@macro amplify_core.storage.storage_path.from_string.example}

///
/// #### Dynamic Path with the current user's IdentityId
/// {@macro amplify_core.storage.storage_path.with_identity_id.example}
class StoragePath {
  /// Creates a [StoragePath] from a static string.
  ///
  /// To create a [StoragePath] from the current user's identity Id, see
  /// [StoragePath.fromIdentityId]
  ///
  /// ### Example
  /// {@template amplify_core.storage.storage_path.from_string.example}
  /// ```
  /// const p = StoragePath.fromString('path/to/object.png');
  /// ```
  /// {@endtemplate}
  const StoragePath.fromString(String path) : _path = path;

  /// {@macro amplify_core.storage.storage_path_from_identity_id}
  ///
  /// ### Example
  /// {@template amplify_core.storage.storage_path.with_identity_id.example}
  /// ```
  /// const p = StoragePath.fromIdentityId((String identityId) => 'users/$identityId/object.png');
  /// ```
  /// {@endtemplate}
  factory StoragePath.fromIdentityId(
    String Function(String identityId) pathBuilder,
  ) =>
      StoragePathFromIdentityId(pathBuilder);

  final String _path;

  @internal
  String resolvePath({String? identityId}) => _path;
}

File: packages/amplify_datastore/ios/Internal/Amplify/Categories/Storage/StoragePath.swift
//
// Copyright Amazon.com Inc. or its affiliates.
// All Rights Reserved.
//
// SPDX-License-Identifier: Apache-2.0
//

import Foundation

public typealias IdentityIDPathResolver = (String) -> String

/// Protocol that provides a closure to resolve the storage path.
///
/// - Tag: StoragePath
public protocol StoragePath {
    associatedtype Input
    var resolve: (Input) -> String { get }
}

public extension StoragePath where Self == StringStoragePath {
    static func fromString(_ path: String) -> Self {
        return StringStoragePath(resolve: { _ in return path })
    }
}

public extension StoragePath where Self == IdentityIDStoragePath {
    static func fromIdentityID(_ identityIdPathResolver: @escaping IdentityIDPathResolver) -> Self {
        return IdentityIDStoragePath(resolve: identityIdPathResolver)
    }
}

/// Conforms to StoragePath protocol.  Provides a storage path based on a string storage path.
///
/// - Tag: StringStoragePath
public struct StringStoragePath: StoragePath {
    public let resolve: (String) -> String
}

/// Conforms to StoragePath protocol.
/// Provides a storage path constructed from an unique identity identifer.
///
/// - Tag: IdentityStoragePath
public struct IdentityIDStoragePath: StoragePath {
    public let resolve: IdentityIDPathResolver
}
